### PR TITLE
Added documentation for remote dropdown. Select2 fixes.

### PR DESCRIPTION
**Pull request by JamieCrisman**
Select2's remote dropdowns mostly already worked, but the documentation is necessary since the remote dropdowns work differently.

I added listeners on select2-focus and select2-blur to wrap ngFocus and ngBlur events, since they were not being fired.

Select2's will no longer close on window scroll. It made infinite scrolling almost impossible as the select would close before the options were loaded.

### CODE REVIEW

**Comment by robmcguinness** [comment]
It seems to me that the Select2 option `query` can better integrate with Angular `$http` services rather than `ajax` option.  If so, this would allow developers to use some of our Availity API factories from within Select2.  This has the added benefit of getting some built in features like async handling from our services.  Here is what I was thinking

```js
// some controller

availity.demo.controller('DropdownController', function(..., avCodesResource) {
  $scope.getCodes = function() {
    return avCodesResource.query();
  };
}
```

```html
<input type="hidden"
  id="dropdownCodes"
  class="form-control select2"
  data-av-dropdown
  data-ng-model="selectedCode"
  data-placeholder="Search for codes"
  data-allow-clear="true"
  data-minimum-input-length="2"
  data-query="getCodes"
/>
```

```js

// AvDropdownController: incomplete example

this.init = function() {
  // ... 

  if(self.options.query) {         
    self.queryFn = self.options.query;
    self.options.query = self.query;
  }
}

// Select2 would fire this function
this.query = function(options) {
  self.queryFn().then(function(response) {
    options.callback({more:false, results: response.data});
  });
};

```

Our Api factories could be enhanced with mixins that could work directly with Select2 + pagination.  

Thoughts?

**Comment by JamieCrisman** [comment]
@robmcguinness I'm trying out what you said, but I'm having a few issues with this.


The code works the same as the ajax call with this code
```javascript
// some controller

availity.demo.controller('DropdownController', function(..., avCodesResource) {
  $scope.getCodes = function() {
    return avCodesResource.query();
  };
}
<input type="hidden"
  id="dropdownCodes"
  class="form-control select2"
  data-av-dropdown
  data-ng-model="selectedCode"
  data-placeholder="Search for codes"
  data-allow-clear="true"
  data-minimum-input-length="2"
  data-query="getCodes"
/>
// AvDropdownController: incomplete example

this.init = function() {
  // ... 

  if(self.options.query) {         
    self.queryFn = self.options.query;
    self.options.query = self.query;
  }
}

// Select2 would fire this function
this.query = function(options) {
  self.queryFn(options).then(function(response) {
    var myResults = response.data.codes;
    if(_.isEmpty(myResults[0].id)){
      _.each(myResults, function(code){
        code.id = code.code;
      });
    }
    var moreVal = (( (response.data.offset/response.data.limit) -1) * 50) < response.data.totalCount;
    options.callback({more: moreVal, results: myResults});
  });
};
```

My issue is that all the customization magic is happening in the avDropdownController. What would be the best way to let the front-end manage the data retrieved by the server before it's put into ``options.callback({..})``?

I'm not quite sure how ``avCodesResource.query()`` is working as well.

**Comment by robmcguinness** [comment]
@JamieCrisman I was hoping the `avCodesResource` module, which is the module that performs the ajax calls, would be the class that transforms the data that can be interpreted by Select2.  Each resource class is derived from `AvApiResource` which has before/after functions for create/update/query that can mutate the data.  Additionally each resource module can have helper methods that can help with pagination.


```js
// AvDropdownController
this.query = function(options) {
  self.queryFn(options).then(function(response) {
    // data should already be in a form that Select2 to can interpret
    options.callback({more: response.more, results: response.results});
  });
};
```

```js
// avCodesResource.js

var AvCodesResourceFactory = function(AvApiResource) {

    var AvCodesResource = function() {
      AvApiResource.call(this, 'codes');
    };

    angular.extend(AvCodesResource.prototype, AvApiResource.prototype, {

      getCodes: function() {

        return this.query().then(function(response) {

          var myResults = response.data.codes;
          if(_.isEmpty(myResults[0].id)){
            _.each(myResults, function(code){
              code.id = code.code;
            });
          }

          var moreVal = (( (response.data.offset/response.data.limit) -1) * 50) < response.data.totalCount;
         
          // send format Select2 can understand
          return {
            more: moreVal,
            results: myResults
          };

        });
      }

    });

    return new AvCodesResource();
  };

  availity.core.factory('avCodesResource', AvCodesResourceFactory);
```

```js
// some controller

availity.demo.controller('DropdownController', function(..., avCodesResource) {
  $scope.getCodes = function() {
    return avCodesResource.getCodes();
  };
}
```

something like that.

File: gulp/config.js [hunk 1]
```diff
@@ -178,6 +178,7 @@ module.exports = {
     src: [
       'node_modules/sinon/pkg/sinon.js',
       'bower_components/jquery/dist/jquery.js',
+      'bower_components/jquery-mockjax/jquery.mockjax.js',
```
Comment (robmcguinness) [inline gulp/config.js]: is mock jax still needed?

File: lib/ui/dropdown/dropdown.js [hunk 2]
```diff
@@ -219,12 +220,31 @@
         });
 
         element.on('change', function(e) {
+          //special case since the ajax handling doesn't bind to the model correctly
+          //this has to do with select2 (v3.5.2) using a hidden field instead of a select for ajax
+          if(!_.isEmpty(avDropdown.options.ajax)) {
```
Comment (robmcguinness) [inline lib/ui/dropdown/dropdown.js]: is this needed since we are using `query` instead of `ajax` now?

**Pull request merged**